frontend/apps/web/components/jobs/NosqlTable/NosqlTable.tsx

```diff
@@ -116,6 +122,50 @@ export default function NosqlTable(props: Props): ReactElement {
 
   const collections = Array.from(Object.keys(schema));
 
+  // useMemo ensures that we don't recreate the set unless the data changes
+  const keySet = useMemo(() => {
+    const set = new Set<string>();
+    data.forEach((item: JobMappingFormValues) => {
+      set.add(`${item.schema}.${item.table}.${item.column}`);
+    });
+    return set;
+  }, [data]);
+
+  // useCallback ensures that we only re-run the function if the keySet changes
+  // we pass in an optional curr value so that we don't throw errors when searching for duplicates and the existing key is there
+  // for ex. if the initial value is "email" and we change it to "emaila" and then back to "email", it throws an error because the key set is made from all of the rows of the data which includes the original value, so we delete it if currValue is valid so we don't throw that error and allow the user to just save their current value without changing it
+  const isDuplicateKey = useCallback(
+    (newValue: string, schema: string, table: string, currValue?: string) => {
+      const key = `${schema}.${table}.${newValue}`;
+      if (currValue) {
+        keySet.delete(`${schema}.${table}.${currValue}`);
+      }
+      return keySet.has(key);
+    },
+    [keySet]
+  );
+
+  // used to calculate the collections that can be updated based on a given key value
+  // for ex. if a collection.key is "a.b.c" and we want to update it to "d.e.c" but "d.e.c" already exists, then we don't want to show the "d.e." collection as an uption for the update
+  const filteredCollections = useCallback(
```

Comment:
a little confused as to what value this is providing.
I read your comment and I thought I understood it, but I couldnt repro what it was trying to accomplish.

I have something like this:

```
collection, key
---------------
col1, foo.bar
col2, foo.bar.baz
```
I was able to edit the first row and update `col1` to be `col2` - if that is what this is trying to prevent?

Reply:
no it's catching that if the collection.key already exists in the data, then we don't want to present that collection to the user as an option that they can switch to.

for ex. say you have this:

collection, key
---------------
col1, foo.bar
col2, foo.bar

and you want to change `col2`, we want to filter out `col1` out of the list of collections that they could change `col2` to so they don't create a duplicate mapping.

But that is another thing we need to consider, and this PR doesn't answer that. This PR prevents duplicate mappings at the leaf level. Do we want to prevent duplicate mappings at any point in the path level? i.e. what you mentioned, since that woudl overwrite the value at `foo.bar.baz` with `foo.bar`

Reply:
ah okay, I think it was this comment `then we don't want to show the "d.e." collection as an uption for the update` that threw me off. 

makes sense.

I think having mappings for `foo.bar` and `foo.bar.baz` is totally fine though.